core/src/main/java/io/aiven/kafka/tieredstorage/transform/RateLimitedInputStream.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2024 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.transform;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+
+import io.github.bucket4j.Bucket;
+
+/**
+ * Input Stream with a limited rate of reading from source.
+ * Aims to provide a stable upload speed that does not affect other streams of work in a broker.
+ */
+public class RateLimitedInputStream extends InputStream {
+    private final InputStream delegated;
+    final Bucket bucket;
+
+    public RateLimitedInputStream(final InputStream delegated, final Bucket bucket) {
+        this.delegated = delegated;
+        this.bucket = bucket;
+    }
+
+    public static Bucket rateLimitBucket(final int uploadRate) {
+        return Bucket.builder()
+            .addLimit(limit ->
+                limit.capacity(uploadRate)
+                    // every 100ms a 10th of the chunk size is added back to bucket
+                    .refillGreedy(uploadRate, Duration.ofSeconds(1)))
+            .build();
+    }
+
+    private void consumeRate(final int rate) {
+        try {
+            bucket.asBlocking().consume(rate);
+        } catch (final InterruptedException e) {
+            throw new RuntimeException("Rate limited consumption of input stream interrupted", e);
+        }
+    }
+
+    @Override
+    public int read() throws IOException {
```

Comment:
I believe the method public int read(byte[] b, int off, int len) should be overridden. Otherwise, public int read(byte[] b, int off, int len) will call read(), which only obtains one token each time, leading to a constant surge in CPU usage.

Reply:
Great catch. I agree we should discard using read() for capacity. Let's only focus on read(byte[] b, int off, int len).
`len` doesn't help much as it's usually the buffer size (8Ki) -- will need to do some compensation after the size read is known. Will draft something for this.

Reply:
I have seen this submission, and it looks good at first glance. I will further test its performance and submit a report here.

Reply:
We can consider adding a test like
```java
    @Test
    void testAllNecessaryCallsAreRateLimited() throws IOException {
        final var buf = new byte[10];

        Bucket bucket = spy(RateLimitedInputStream.rateLimitBucket(1));
        var rateLimitedInputStream = new RateLimitedInputStream(new ByteArrayInputStream(new byte[100]), bucket);
        rateLimitedInputStream.read();
        verify(bucket).asBlocking();

        bucket = spy(RateLimitedInputStream.rateLimitBucket(1));
        rateLimitedInputStream = new RateLimitedInputStream(new ByteArrayInputStream(new byte[100]), bucket);
        rateLimitedInputStream.read(buf);
        verify(bucket).asBlocking();

        bucket = spy(RateLimitedInputStream.rateLimitBucket(1));
        rateLimitedInputStream = new RateLimitedInputStream(new ByteArrayInputStream(new byte[100]), bucket);
        rateLimitedInputStream.read(buf, 0, 10);
        verify(bucket).asBlocking();

        bucket = spy(RateLimitedInputStream.rateLimitBucket(1));
        rateLimitedInputStream = new RateLimitedInputStream(new ByteArrayInputStream(new byte[100]), bucket);
        rateLimitedInputStream.readNBytes(10);
        verify(bucket).asBlocking();

        bucket = spy(RateLimitedInputStream.rateLimitBucket(1));
        rateLimitedInputStream = new RateLimitedInputStream(new ByteArrayInputStream(new byte[100]), bucket);
        rateLimitedInputStream.readNBytes(buf, 0, 10);
        verify(bucket).asBlocking();

        bucket = spy(RateLimitedInputStream.rateLimitBucket(1));
        rateLimitedInputStream = new RateLimitedInputStream(new ByteArrayInputStream(new byte[100]), bucket);
        rateLimitedInputStream.readAllBytes();
        verify(bucket, atLeastOnce()).asBlocking();
    }
```
which checks that all the relevant methods when called interact with the bucket. This means these calls go through our implementation, which we know how it's done.

Reply:
@ivanyu good call, adding this test.